### PR TITLE
Block Supports: Allow skipping serialization of font size

### PR DESCRIPTION
**Pull request by ockham**
## Description
Part of #28913. Counterpart to #29142+#29253, or #30035.

Find affected blocks by grepping `"fontSize": true` in `block.json` files.

## How has this been tested?

(Based on @nosolosw's https://github.com/WordPress/gutenberg/pull/30035#issuecomment-804218854, and adapted for fontSize block support.)

- Install and activate the TT1-blocks theme.
- Add `"__experimentalSkipFontSizeSerialization": true` as a new key-value pair within the `supports` object for any block that also has `"fontSize": true` set. In the following, we'll be using the Heading block.

Here's a patch for your convenience:

```diff
diff --git a/packages/block-library/src/heading/block.json b/packages/block-library/src/heading/block.json
index 8d7e0fdd5c..a35cf68b6c 100644
--- a/packages/block-library/src/heading/block.json
+++ b/packages/block-library/src/heading/block.json
@@ -74,6 +74,7 @@
                                }
                        }
                },
+               "__experimentalSkipFontSizeSerialization": true,
                "__unstablePasteTextInline": true
        },
        "editorStyle": "wp-block-heading-editor",
```

The expected result is that:

- The UI control for font size is shown.
- Using the UI control still syncs the block attributes.
	- Use the font size control.
	- Use the dropdown to select a "named" font size, e.g. "Large".
	- Open the Code Editor and verify that the block has the proper font size value as `fontSize` attribute, e.g. `<!-- wp:heading {"fontSize":"large"} -->` -- but _not_ as inlined markup (HTML `class="has-large-font-size"` attribute on the `<h2 />` tag).
	- Return to the Visual Editor.
	- Use the font size control to set the block's font size to a numeric value (e.g. 25px).
	- Open the Code Editor and verify that the block has the proper font size value as a nested `typography:{ fontSize: ... } }` attribute, e.g. `
` -- but _not_ as inlined markup (HTML `style="font-size:25px"` attribute on the `<h2 />` tag).
	- Make sure to also change the line height control, to verify that it is unaffected (i.e. there's still a `style="line-height:...` attribute on the `<h2 />` tag).

### Before

font-size-before

### After

font-size

## Question

`fontSize` is a non-experimental, "flat" (non-array) block-supports flag. This is the reason why we're introducing _another_ "flat" flag to skip serialization specifically of font size here: `__experimentalSkipFontSizeSerialization`. 

Compare that to "nested" block-supports flags (such as `color`), where we can include a "generic" `__experimentalSkipSerialization` in that array: https://github.com/WordPress/gutenberg/blob/aa272f8432ca055961005a62c200a856e8017cb7/packages/block-library/src/button/block.json#L56-L59

A dedicated `__experimentalSkipFontSizeSerialization` flag doesn't scale particularly well, and might get awkward once we allow skipping of serialization for other typographic properties (line height, font family, font appearance, text decoration, text transform). Will each of these get their own `__experimentalSkip...Serialization`? Or will we have to group typographic block-supports flags into a shared `typography` object first? (But that might affect themes that use these already, won't it?) 🤔

### CODE REVIEW

**Review by aaronrobertshaw — COMMENTED** [review]
This tests as advertised and is looking pretty good 👍 

I believe we might need an additional check for skipping serialization in `withFontSizeInlineStyles` in `hooks/font-size.js`. 

This will prevent the inline styles being applied to the block when selecting a different font size from the dropdown or a custom font size that matches a preset.

```js
	if (
		! hasBlockSupport( blockName, FONT_SIZE_SUPPORT_KEY ) ||
		hasBlockSupport(
			blockName,
			'__experimentalSkipFontSizeSerialization'
		)
	) {
		return <BlockListBlock { ...props } />;
	}
```

Regarding the issue of the flat `__experimentalSkipFontSizeSerialization` flag, it would feel neater if the typography support flags were collected under `typography` like `color` or `border`. As mentioned though that would be a breaking change for themes already using them.

Could an intermediate step be to use a `typography.__experimentalSkipSerialization`  or `__experimentalSkipTypographySerialization` flag and check that one flag across the typography related block supports? 

In other words do we need separate `__experimentalSkip...Serialization` flags?

**Review by jorgefilipecosta — COMMENTED** [review]
> Could an intermediate step be to use a typography.__experimentalSkipSerialization

I like the idea of this intermediate step.


Our block.json supports API right now has inconsistencies in how we handle typography vs colors:

```
		"color": {
			"gradients": true,
			"link": true
		},
		"fontSize": true,
		"lineHeight": true,
```
It is also inconsistent with theme.json given that there we have:
```
			"typography": {
				"fontSize": "value",
				"lineHeight": "value",
			}
```

The intermediate step allows us not to build on these inconsistencies. I guess we can change the shape of block.json and be back-compatible. We can have a back-compatibility layer that checks if things like fontSize and lineHeight are present in the data structure and if yes nests them under typography. Our code excluding the back-compatibility layer would only rely on the new structure.
The shape change would be done in another PR. I'm not sure if a change like this would require a new "apiVersion", I guess it is something minor that does not requires it, but I'm cc'ing @gziolo in case he has some thoughts.

**Comment by gziolo** [comment]
> The shape change would be done in another PR. I'm not sure if a change like this would require a new "apiVersion", I guess it is something minor that does not requires it, but I'm cc'ing @gziolo in case he has some thoughts.

It's already document at https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/, but it would be good to validate whether it was released in the previous major version of WordPress. If not then we can simply change it and keep the backward compatibility layer only in the plugin for a few more releases.

In general, the question goes to @nosolosw how much `supports` in `block.json` should align with `theme.json` as it doesn't necessarily translate one to one. Conceptually, I would be in favor of using the same naming and grouping in both places to make it consistent. If they are also grouped in the UI in the same way, then it adds another reason to make the extra efforts to bring harmony how all that fits together.

Concluding, if this is only a simple mapping layer that would have to map some properties from their deprecated names, then we should definitely go for it. In the documentation, we should simply remove old variants in favor of the refined approach.

**Comment by ockham** [comment]
> This tests as advertised and is looking pretty good 👍

Yay! 🎉 Thanks for reviewing and testing!

> I believe we might need an additional check for skipping serialization in `withFontSizeInlineStyles` in `hooks/font-size.js`.
>
> This will prevent the inline styles being applied to the block when selecting a different font size from the dropdown or a custom font size that matches a preset.
> 
> ```js
> 	if (
> 		! hasBlockSupport( blockName, FONT_SIZE_SUPPORT_KEY ) ||
> 		hasBlockSupport(
> 			blockName,
> 			'__experimentalSkipFontSizeSerialization'
> 		)
> 	) {
> 		return <BlockListBlock { ...props } />;
> 	}
> ```

I stared at this file for a while when I was working on this PR and decided to leave it as-it. I don't claim I understood it fully at the time (and there's a chance that I still don't 😅 ), but my reading was as follows:
This only modifies the [`BlockListBlock` wrapper](https://developer.wordpress.org/block-editor/reference-guides/filters/block-filters/#editor-blocklistblock) -- which is editor-specific, and doesn't get serialized.

The logic in `withFontSizeInlineStyles` will only kick in if the `fontSize` attribute is set (to a "named" font size, such as `large`), whereas `style.typography.fontSize` isn't. In that case, the logic will look up the font size (in px) per information found in `theme.json`'s `typography.fontSizes` setting. It will then add that `fontSize` to the `BlockListBlock` wrapper's `style`.

I think it's fine to retain this behavior, even if we're disabling serialization 🤔 The only thing that changes compared to the previous situation is that we're no longer attaching a `class` or `style` to the actual markup element. But we're still storing the `fontSize` attribute; so I think whatever necessitated the style to be added to the `BlockListBlock` wrapper still holds.

LMK if this makes sense -- I'm still not 100% sure myself 😅 

This comment is a bit cryptic: https://github.com/WordPress/gutenberg/blob/928e6a46e5543bbf52489479d3e3bb6eaadc4b18/packages/block-editor/src/hooks/font-size.js#L156-L157 -- Not sure _how_ a theme would load font-size classes in the editor? I thought it would do that through the `typography.fontSizes` setting -- but that's what `withFontSizeInlineStyles` is evaluating anyway 🤔  cc/ @nosolosw @jorgefilipecosta

**Comment by ockham** [comment]
> > Could an intermediate step be to use a typography.__experimentalSkipSerialization
> 
> I like the idea of this intermediate step.

Sounds good to me, too 👍 

> Our block.json supports API right now has inconsistencies in how we handle typography vs colors:
> 
> ```
> 		"color": {
> 			"gradients": true,
> 			"link": true
> 		},
> 		"fontSize": true,
> 		"lineHeight": true,
> ```
> 
> It is also inconsistent with theme.json given that there we have:
> 
> ```
> 			"typography": {
> 				"fontSize": "value",
> 				"lineHeight": "value",
> 			}
> ```
>
> The intermediate step allows us not to build on these inconsistencies. I guess we can change the shape of block.json and be back-compatible. We can have a back-compatibility layer that checks if things like fontSize and lineHeight are present in the data structure and if yes nests them under typography. 

We already have something like that in the [`useEditorFeature` hook](https://github.com/WordPress/gutenberg/blob/928e6a46e5543bbf52489479d3e3bb6eaadc4b18/packages/block-editor/src/components/use-editor-feature/index.js), don't we? Except it almost looks like we're doing the opposite there -- e.g. mapping a nested `typography.customLineHeight` to a flat `enableCustomLineHeight` 🤔 

https://github.com/WordPress/gutenberg/blob/928e6a46e5543bbf52489479d3e3bb6eaadc4b18/packages/block-editor/src/components/use-editor-feature/index.js#L37-L38

This is getting pretty confusing 😅 All said and done, would we end up with
- a editor-feature deprecation that transforms `typography.customLineHeight` into `enableCustomLineHeight` (as we already do), and
- a theme-setting deprecation that transforms `lineHeight` into `typography.lineHeight`? Oh boy 😬 

> Our code excluding the back-compatibility layer would only rely on the new structure.
> The shape change would be done in another PR. I'm not sure if a change like this would require a new "apiVersion", I guess it is something minor that does not requires it, but I'm cc'ing @gziolo in case he has some thoughts.

Should this PR be blocked by the shape change, or can we do that after merging this one?

**Comment by gziolo** [comment]
> Should this PR be blocked by the shape change, or can we do that after merging this one?

Definitely not a blocker. It should be done separately so we could reason about all changes in one place.

**Comment by aaronrobertshaw** [comment]
> LMK if this makes sense -- I'm still not 100% sure myself 😅

Thanks for the extra background @ockham, I don't claim to be an expert myself either 😄 

> I stared at this file for a while when I was working on this PR and decided to leave it as-it. I don't claim I understood it fully at the time (and there's a chance that I still don't 😅 ), but my reading was as follows:
This only modifies the BlockListBlock wrapper -- which is editor-specific, and doesn't get serialized.

While it's correct that it's editor specific, the font size styles are serialized for the block's actual markup in the DOM. The markup you see in the code editor view is untouched. 

So changing the font size selection from the drop down would impact the visual size of the text. If this is being used on a more complicated block where the font size should only be applied to an inner element, while it is in the editor, this rogue style would cause a disjoint between what the user sees in the editor vs frontend.

| Current Situation | Updated `withFontSizeInlineStyles` |
|---|---|
| FontSizeSerialization | FontSizeSerialization2 |

With the tweak to `withFontSizeInlineStyles` the styles applied to the block's markup in the DOM more closely matches the markup seen in the code editor view as well.


> Should this PR be blocked by the shape change, or can we do that after merging this one?

A separate PR for the shape change gets my vote as well and shouldn't block this one.

**Comment by ockham** [comment]
Thanks for taking the time to explain, and to prepare those Gifs, @aaronrobertshaw!

> So changing the font size selection from the drop down would impact the visual size of the text. If this is being used on a more complicated block where the font size should only be applied to an inner element, while it is in the editor, this rogue style would cause a disjoint between what the user sees in the editor vs frontend.

This is where the penny dropped for me. Thanks, I'll apply the change you suggested in https://github.com/WordPress/gutenberg/pull/30879#pullrequestreview-640688330 😊 

>> Should this PR be blocked by the shape change, or can we do that after merging this one?
> A separate PR for the shape change gets my vote as well and shouldn't block this one.

Glad that you and @gziolo agree! ✨

**Comment by ockham** [comment]
> > So changing the font size selection from the drop down would impact the visual size of the text. If this is being used on a more complicated block where the font size should only be applied to an inner element, while it is in the editor, this rogue style would cause a disjoint between what the user sees in the editor vs frontend.
> 
> This is where the penny dropped for me. Thanks, I'll apply the change you suggested in [#30879 (review)](https://github.com/WordPress/gutenberg/pull/30879#pullrequestreview-640688330) 😊

Done per the last couple of commits. Should be ready for another look 😊

File: lib/block-supports/typography.php
Comment (aaronrobertshaw) [inline lib/block-supports/typography.php]: The `__experimentalSkipTypographySerialization` flag here doesn't match the `__experimentalSkipFontSizeSerialization` one used in the client-side checks. 

```suggestion
	$skip_font_size_support_serialization = _wp_array_get( $block_type->supports, array( '__experimentalSkipFontSizeSerialization' ), false );
```

**Review by aaronrobertshaw — CHANGES_REQUESTED** [review]
Thanks for the updates on this 👍 

I'd say this one is nearly ready to merge. I just have one comment relating to a difference in flags being checked server-side vs client.

The skipping of font size serialization works well for static blocks. The issue mentioned above however means you'd need a second flag for the same behaviour to occur for a dynamic block rendering via PHP.

Adding the `__experimentalSkipFontSizeSerialization` flag to the Post Tags block results in the experience below i.e. serialization still occurs for dynamic blocks on the frontend. Bringing the flag checked within `typography.php` in line fixes this issue.

File: lib/block-supports/typography.php
Comment (ockham) [reply lib/block-supports/typography.php]: Ah, good spot, thanks! Silly oversight on my part, since I started this PR by copying some stuff from #30880 😅

**Review by aaronrobertshaw — APPROVED** [review]
LGTM ✨ 

This tests well for me now with static and dynamic blocks in both the editor and on the frontend.

**Comment by ockham** [comment]
Thanks @aaronrobertshaw! Really appreciated your review! ✨

**Pull request merged**
